include/deal.II/matrix_free/assemble_operator.h

```diff
@@ -0,0 +1,278 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2014 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#pragma once
+
+#include <deal.II/lac/vector.h>
+
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/graph_selectors.hpp>
+#include <boost/graph/minimum_degree_ordering.hpp>
+#include <boost/graph/sequential_vertex_coloring.hpp>
+#include <boost/graph/smallest_last_ordering.hpp>
+#include <boost/property_map/property_map.hpp>
+#include <boost/property_map/shared_array_property_map.hpp>
+
+#include <set>
+#include <vector>
+
+// TODO distributed graph
+
+namespace dealii
+{
+  namespace MatrixFreeUtilities
+  {
+    // Typedefs for boost graph types.
+    namespace
+    {
+      using Graph =
+        boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS>;
+      using vertex_descriptor  = boost::graph_traits<Graph>::vertex_descriptor;
+      using vertices_size_type = boost::graph_traits<Graph>::vertices_size_type;
+      using vertex_index_map =
+        boost::property_map<Graph, boost::vertex_index_t>::const_type;
+      using Edge = std::pair<vertices_size_type, vertices_size_type>;
+    } // namespace
+
+    /**
+     * Cache the graph generated from a sparsity pattern for later computations.
+     * This is useful if the matrix does not change its pattern.
+     * // TODO cache entries
+     */
+    struct GraphCache
+    {
+      Graph                           graph;
+      std::vector<vertices_size_type> color_vec;
+      boost::iterator_property_map<vertices_size_type *, vertex_index_map>
+                                                   coloring;
+      vertices_size_type                           num_colors;
+      std::vector<std::vector<vertices_size_type>> entries_per_color;
```

Comment:
What about this?

Reply:
Missed that. Whats the benefit of using that storage scheme? It feels like a lot more space for errors.
If we precompute the indices (as indicated in another comment), this would become obsolete.